### PR TITLE
Add spinlock and switch nodelock to use it

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

More code towards nodegroup elements

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Spinlock.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+class Spinlock {
+ public:
+  Spinlock() = default;
+  Spinlock(const Spinlock&) = delete;
+  Spinlock& operator=(const Spinlock&) = delete;
+  Spinlock(Spinlock&&) = delete;
+  Spinlock& operator=(Spinlock&&) = delete;
+  ~Spinlock() = default;
+
+  /// \brief Acquire the lock. Will block until the lock is acquire.
```
Comment (kidder) [inline src/Parallel/Spinlock.hpp]: acquired

File: src/Parallel/Spinlock.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  void lock() {
+    for (;;) {
+      // Optimistically assume the lock is free on the first try
+      if (not lock_.exchange(true, std::memory_order_acquire)) {
+        return;
+      }
+      // Wait for lock to be released without generating cache misses
+      while (lock_.load(std::memory_order_relaxed)) {
+        // Issue X86 PAUSE or ARM YIELD instruction to reduce contention between
```
Comment (kidder) [inline src/Parallel/Spinlock.hpp:32]: is there a reason to keep this comment?

File: src/Parallel/NodeLock.cpp [hunk 3]
```diff
@@ -5,18 +5,23 @@
...
 #include "Utilities/Gsl.hpp"
 
 namespace Parallel {
 
-NodeLock::NodeLock() : lock_(std::make_unique<CmiNodeLock>(CmiCreateLock())) {}
+NodeLock::NodeLock()
+    : lock_(std::make_unique<typename decltype(lock_)::element_type>())
 
-NodeLock::NodeLock(NodeLock&& moved_lock) : lock_(std::move(moved_lock.lock_)) {
+{}
```
Comment (kidder) [inline src/Parallel/NodeLock.cpp]: something went wrong with the formatting...

File: src/Parallel/NodeLock.cpp [hunk 4]
```diff
@@ -5,18 +5,23 @@
 
 #include <converse.h>
```
Comment (kidder) [inline src/Parallel/NodeLock.cpp]: remove?

File: src/Parallel/NodeLock.hpp [hunk 5]
```diff
@@ -8,6 +8,8 @@
 #include <memory>
 #include <pup.h>
 
+#include "Parallel/Spinlock.hpp"
```
Comment (kidder) [inline src/Parallel/NodeLock.hpp:9]: rm converse.h include?

File: tools/FileTestDefs.sh [hunk 6]
```diff
@@ -622,7 +622,8 @@ standard_checks+=(enable_if)
 # Check for noexcept
 noexcept() {
     whitelist "$1" \
-              'src/Parallel/StaticSpscQueue.hpp' && \
+              'src/Parallel/StaticSpscQueue.hpp' \
+              "src/Parallel/NodeLock..pp$" && \
```
Comment (kidder) [inline tools/FileTestDefs.sh:626]: for my education, what does the `$` at the end of the string do?

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash the changes

File: src/Parallel/Spinlock.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Parallel/Spinlock.hpp:32]: I've added a comment explaining that this is something we might want to experiment with in the future, so I'm keeping it here as "knowledge"

File: tools/FileTestDefs.sh (hunk 6 above)
Comment (nilsdeppe) [reply tools/FileTestDefs.sh:626]: That's end-of-line in regex

**Comment by nilsdeppe** [comment]
I squashed in the few fixes into the commits. Thanks for the review :)

**Review by kidder — APPROVED** [review]

File: src/Parallel/NodeLock.cpp [hunk 7]
```diff
@@ -3,20 +3,25 @@
...
 
+#include "Parallel/Spinlock.hpp"
 #include "Utilities/ErrorHandling/Error.hpp"
 #include "Utilities/Gsl.hpp"
 
 namespace Parallel {
 
-NodeLock::NodeLock() : lock_(std::make_unique<CmiNodeLock>(CmiCreateLock())) {}
+NodeLock::NodeLock()
+    : lock_(std::make_unique<typename decltype(lock_)::element_type>()) {}
```
Comment (knelli2) [inline src/Parallel/NodeLock.cpp:15]: [optional] Why go through the unique_ptr to get the type? Just use `SpinLock`? Same below.

**Review by knelli2 — APPROVED** [review]

File: src/Parallel/NodeLock.cpp (hunk 7 above)
Comment (nilsdeppe) [reply src/Parallel/NodeLock.cpp:15]: This makes it easier to switch the type if we want to in the future. I was going back and forth a bit trying to figure out what's best, including std::mutex, so only having to switch the type in the member variable was easier.

**Pull request merged**